clang-tools-extra/clangd/ModulesBuilder.cpp

```diff
@@ -127,33 +127,39 @@ struct ModuleFile {
   std::string ModuleFilePath;
 };
 
-bool IsModuleFileUpToDate(PathRef ModuleFilePath,
-                          const PrerequisiteModules &RequisiteModules,
-                          llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> VFS) {
+bool IsModuleFileUpToDate(
+    PathRef ModuleFilePath,
+    const PrerequisiteModules &RequisiteModules,
+    const CompilerInvocation &CI,
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> VFS) {
+  CompilerInstance Clang;
```

Comment:
i think you can move all of this CompilerInstance creation logic into `IsModuleFilesUpToDate` and just pass the ASTReader here. that way we can save up a bunch of work/allocations.

Reply:
this comment still applies

Reply:
We may not be able to do it since ASTReader is stateful. If the ASTReader reads multiple module files (BMIs), the ASTReader will check automatically if these module files conflicts with each other. It doesn't matter in an ideal world. But it is common that users wrote invalid codes especially during the process of editing.